### DISCUSSION TITLE
Example for custom references storing additional information (is this how langium is intended to be used?)

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#615] [Q&A] [answered]
Thank you for the nice documentation (https://langium.org/docs/). I really appreciated the very clear description of the document lifecycle!
I made some experiments with **custom references storing additional information**: https://github.com/goto40/msg-dsl-mixed-lookup-and-path 
I use langium 0.4.0.

```
//...
MsgArrayAttribute: 'array' name=ID ':' msg_type=[MsgType:FQN] '[' dim=[MsgArrayDim:FQN] ']';
                                                                // ^^^^^^^^^^^^^^^^^^^
interface MsgScalarAttributeRef {   // <---------------------
    path: MsgScalarAttribute[]
}

type MsgArrayDim = MsgConst|MsgScalarAttributeRef;
//...
```

I created a scope provider which creates elements of type `MsgScalarAttributeRef` and **I would like to know if this is how it is intended to use langium** (my solution works for me).
```
                const element: MsgScalarAttributeRef = {
                    $type: "MsgScalarAttributeRef",
                    path: entry[1]
                };
                // ...
               this.descriptionProvider.createDescription(
                    element, entry[0], getDocument(element.path[element.path.length-1])
                );
```
I am **concerned about the reference to the document**: I only link my reference to one document (`element.path[element.path.length-1])`), although I have elements from
more than one document involved in my reference.

**Comment by msujew** [comment, marked as answer]
Hey @goto40,

You're just trying to reference a single element using it's fully qualified name, right? Your approach is a bit unconventional/overengineered, since what you're trying to do can be relatively easily accomplished using the [scoping in our domain-model example](https://github.com/langium/langium/blob/main/examples/domainmodel/src/language-server/domain-model-scope.ts). Although your fully qualified name is a chain of elements, it actually only references a single element which is at the end of the chain. That's exactly what the domain-model scoping does.

Note that your approach using synthetic AST nodes also breaks a lot of LSP behavior, notably the go to definition and find references feature.

**Reply by goto40** [reply]
I think (but probably I am wrong) the use case I illustrated is more complicated than in the domain example:

My lookup for attributes, like `header.n` (or in the example below `y.z.i`) involves other scope resolutions, because we have to evaluate the (referenced) type of `header` to compute `n` as a possible resolution target). The resolved element is not one element, but a path to one element (possibly involving multiple files).

X.item
```
package x {
    rawtype int
    package constants {
        const N = 100
    }
    struct X {
        scalar y: y.Y
        array a: int[y.z.i]
        array b: int[constants.N]
    }
}
```

Y.item
```
package y {
    struct Y {
        scalar z: z.Z
    }
}
```

Z.item
```
package z {
    rawtype int
    struct Z {
        scalar i: int
    }
}
```

As I said my solution works for me - even when I "break" and rename the package y: then (while editing) my reference to `y.z.i` immediately gets marked as erroneous. This happens even that my reference is linked to Z.item (and not Y.item) - I do not understand why but it works.

An alternative way is to lookup all individual parts of the name (`header` and then `n`, recursively) - something like `Ref: ref=[MsgScalarAttribute] ('.' next=Ref)?;`, but this will not go well with the alternative `FQN`-lookup for `const`.

I am happy to discuss this.
I would be happy to learn how this can be achieved in a better way...

**Reply by msujew** [reply]
> I think (but probably I am wrong) the use case I illustrated is more complicated than in the domain example

Honestly, it sounds like the exact same use case. Under the hood both the domain model example and your code are doing the same thing. You expose an object of your AST using a fully qualified name. In theory you can create a real chain of references (i.e. first create a reference to `header`, then create a reference to `n`), but that's not what you're doing. `[header.n]` creates only one reference, since `header.n` is one of the names of `n` that is exposed to the scope. The same thing is happening in the domain-model example.

To actually create a chain of references you would need to do something like this:

```
MemberCall: ref=[MsgScalarAttribute] ('.' next=MemberCall)?;
```

That will create a linked list of `MemberCall` elements, but the scoping for that is more complicated, since it requires to resolve the parent reference first, before being able to resolve child references.

> As I said my solution works for me - even when I "break" and rename the package y: then (while editing) my reference to y.z.i immediately gets marked as erroneous. This happens even that my reference is linked to Z.item (and not Y.item) - I do not understand why but it works.

Right, it makes perfect sense that it works, since you're doing exactly what we're doing in the domain-model scoping, but in a more complicated manner. I'm not saying it doesn't work, it's just that you can make it work with less effort using a more sophisticated and smarter scoping design.

**Reply by goto40** [reply]
Thank you for the comment. I will have to think about it. To me it still looks different - but probably I misinterpret something. I will re-comment later...

**Reply by goto40** [reply]
Thank you very much for the information!

I am aware of the `MemberCall` solution you mentioned. I used this in other places. However, I was not able to "mix" 
  * the `MemberCall`-scope resolution (which **resolves each part of a dot-separated name individually** to produce the path I need) 
  * with the normal scope resolution (in my example the `const`-lookup implemented in my `MessageDslDescriptionProvider` and `MessageDslScopeComputation` (which resolves the **whole FQN to the one `const`-element**).

I "mix" both lookup here: `dim=[MsgArrayDim:FQN] `. `dim` can either be a FQN to a `const` element (simple lookup like in the domain model example) - or `dim` can be a path through the model to navigate to a specific member (like in your `MemberCall` example).

I do not want to tune the grammar by introducing a keyword separating both kinds of references (then the problem is solved, because I do not need to mix both lookup logics.

Of course I could resolve the scope to the `const`-element with the same technique as for the `MemberCall` (resolve each package name individually until I reach a `const`element) - but this would be quite ineffective (I would not use the indexed information any more).

This is what I mean with **mixed lookup** in https://github.com/goto40/msg-dsl-mixed-lookup-and-path.
I was quite **happy with my solution there**, but my feeling was that it is not using langium as it is meant to be used (what you confirmed above): I was able to use a dot-separated name for a reference which produced the path to the element...

Did you encounter a similar problem in the past? How was it solved.
I would be happy to provide more information if my problem is not clear enough.

(Or do you still think I have just reproduced the domain language example? In that case which reference corresponds to my `header.n` Link?)

**Reply by goto40** [reply]
> You're just trying to reference a single element using it's fully qualified name, right? 

No I am trying to reference a single element and the path to it (like in your `MemberCall` example above).

**Reply by goto40** [reply]
Maybe the title of this discussion is inappropriate: the "mix" aspect mentioned in the last comments is not reflected there...

**Reply by msujew** [reply]
Nevermind, you're actually right. I didn't grok your code correctly. It is more like the `MemberCall` example, but you're using a single FQN and a synthetic element list instead of following a chain of real references. That makes sense, though I wouldn't recommend. Like I said, it breaks a few LSP services. For example, you cannot use the "go to definition" or "find references" feature, as the references aren't "real".

**Reply by goto40** [reply]
Thank you for the fast response!

**Reply by goto40** [reply]
>> Note that your approach using synthetic AST nodes also breaks a lot of LSP behavior, notably the go to definition and find references feature.

I just updated my scope provider: The synthetic AST node now "mimics" the final target (the last part of my path) by copying its contents:
```
                const element: MsgScalarAttributeRef = {
                    ...path[path.length-1],          // copy target AST node
                    $type: MsgScalarAttributeRef,    // type of new AST node
                    path: path                       // path to AST node
                };
                // now: return the new element into the current `Scope`
```
With this the "go to definition" feature seems to work - but not the "find references".
 * Is the AST node used as key somewhere? My solution "mimics" the final target by copying its contents, but the new synthetic node is a different object than the final target.
 * Any idea where to start to allow the find references feature...?

Notes:
 * Since I store the path to my destination I wonder if there is a relation to #236 (maybe a solution to #236 will make my solution obsolete).
 * I also tried the solution illustrated with the `MemberCall`above: The problem is that I also have to lookup my constants the same way, but I ran into problems when using the same package name in different files (the wrong package was selected...). Maybe I start another discussion for that if I continue investigating.

**Reply by goto40** [reply]
I still use this approach here, since I do not see how resolve my problem here.

I have references of the form `myref=[MyType:FQN]`, where I want to resolve from [qualified name scoping](https://langium.org/guides/scoping/qualified-name/) *and* [class member scoping](https://langium.org/guides/scoping/class-member/).

My solution (see above: **returning a synthetic AST element** in my scope provider) "works" for me. 
**My main concern is**: how probable is that my "solution" **breaks with future version of langium?**

**Reply by msujew** [reply]
> My main concern is: how probable is that my "solution" breaks with future version of langium?

Very unlikely. We're quite happy with the current scoping infrastructure, and even though we might extend on it (see https://github.com/langium/langium/pull/1121), we don't plan on changing the APIs.
